crates/walrus-service/src/node/storage/blob_info.rs

```diff
@@ -1273,40 +1273,57 @@ impl Mergeable for BlobInfoV1 {
     }
 
     fn merge_new(operand: Self::MergeOperand) -> Option<Self> {
-        let BlobInfoMergeOperand::ChangeStatus {
-            change_type: BlobStatusChangeType::Register,
-            change_info:
-                BlobStatusChangeInfo {
-                    deletable,
-                    epoch: _,
-                    end_epoch,
-                    status_event,
-                    blob_id: _,
-                },
-        } = operand
-        else {
-            tracing::error!(
-                ?operand,
-                "encountered an update other than 'register' for an untracked blob ID"
-            );
-            return None;
-        };
-
-        Some(
-            if deletable {
-                ValidBlobInfoV1 {
-                    count_deletable_total: 1,
-                    latest_seen_deletable_registered_epoch: Some(end_epoch),
-                    ..Default::default()
+        match operand {
+            BlobInfoMergeOperand::ChangeStatus {
+                change_type: BlobStatusChangeType::Register,
+                change_info:
+                    BlobStatusChangeInfo {
+                        deletable,
+                        epoch: _,
+                        end_epoch,
+                        status_event,
+                        blob_id: _,
+                    },
+            } => Some(
+                if deletable {
+                    ValidBlobInfoV1 {
+                        count_deletable_total: 1,
+                        latest_seen_deletable_registered_epoch: Some(end_epoch),
+                        ..Default::default()
+                    }
+                } else {
+                    ValidBlobInfoV1 {
+                        permanent_total: Some(PermanentBlobInfoV1::new_first(
+                            end_epoch,
+                            status_event,
+                        )),
+                        ..Default::default()
+                    }
                 }
-            } else {
+                .into(),
+            ),
+            BlobInfoMergeOperand::MarkInvalid {
+                epoch,
+                status_event,
+            } => Some(BlobInfoV1::Invalid {
+                epoch,
+                event: status_event,
+            }),
+            BlobInfoMergeOperand::MarkMetadataStored(_) => Some(
```

Comment:
Question: why do we need this case? It seems that MarkMetadataStored should always be triggered when blob is registered (otherwise, the op that stores blob metadata won't be executed). Or do I miss something?

Reply:
We update the metadata stored to `false` [when deleting the metadata](https://github.com/MystenLabs/walrus/blob/9b2aa5600c4d4a96e1d0f5c4723cacee3101f4ff/crates/walrus-service/src/node/storage.rs#L662), which in turn [is called when marking a blob as invalid](https://github.com/MystenLabs/walrus/blob/9b2aa5600c4d4a96e1d0f5c4723cacee3101f4ff/crates/walrus-service/src/node/blob_event_processor.rs#L201). 

However, now that I think about it, we can skip this blob-info update as it has no effect after having already set it to invalid.

Reply:
> we can skip this blob-info update as it has no effect after having already set it to invalid.

Yeah, this was part of my original thoughts too. To me, it feels that if merge_new receiving MarkMetadataStored(), it indicates a bug, since this shouldn't happen. If my understanding is correct, I wonder if we should remove this branch to not overlook a potential issue. But I agree that this may not be super important based on the current code.

Also, it seems that we should set `is_metadata_stored` to the value passed in BlobInfoMergeOperand::MarkMetadataStored()?

Reply:
+1 to @halfprice 's comment about removing the `MarkMetadataStored` branch as it seems logically impossible.

Reply:
Ok, I've now changed the behavior for the `MarkMetadataStored` case to what we had before.